design/pipeline-artifacts.md

```diff
@@ -83,6 +144,32 @@ Or to avoid downloading any of the artifacts at all:
 
 You can add files to any artifact multiple times in the same job, in different jobs in the same stage, and in different stages in the same pipeline, or you can `seal` an artifact at any time to prevent further files from being added.
 
+## Variables
+
+In addition to the new `$(Pipeline.Workspace)` variable, we introduce a variable per resource indicating where it was checked out on disk. For example:
+```yaml
+resources:
+  pipelines:
+  - pipeline: foo
+
+jobs:
+- job: maker
+  steps:
+  - script: ./build.sh
+  - upload: dist/**/*
+
+- job: consumer
+  dependsOn: maker
+  steps:
+  - download:   # remember, blank indicates we download all artifacts from all pipelines
+  - bash: |
+      echo $(Pipeline.Resources.foo)   # pipeline variable syntax
+      echo $PIPELINE_RESOURCES_FOO     # environment variable syntax
```

Comment:
Per our conversation earlier.  I wonder if we should choose to stop propagating everything into the environment variables and simply rely on either macro replacement in tasks or mapping using the ENV block for scripts.  There is a downside as it will make things a bit more complex in some cases for scripts run from the repo rather than inline.  There also could be a downside for task authors as there is now no way to get all of the artifact info from variables without the user explicitly giving you that input but I am not sure that is a requirement.

Reply:
I think that having environment variables provides awesome flexibility for people to build their own stuff on top of what we provide. For example if someone wants to do some scanning on the artifacts before they deployment they can write a script which picks up all the artifact paths based on a prefix and feeds it into their scanning tools.

Is there a reason not to propagate in this particular case, it seems like it'll be helpful in some scenarios.